File: ports/dotnet/src/RussianDocs.DocumentProcessing/Imaging/Crop.cs
using OpenCvSharp;

namespace RussianDocs.DocumentProcessing.Imaging;

/// <summary>
/// Cropping. **The only sanctioned crop path in the port.**
///
/// <para>
/// This is the single most dangerous divergence in the whole exercise, and it is worth spelling out
/// because the failure mode is a crash on a rare input rather than a wrong number on a common one.
/// </para>
///
/// <para>
/// <c>img[y1:y2, x1:x2]</c> in Python does not validate: an upper bound past the edge is silently
/// CLAMPED, and a negative start counts from the end. <c>new Mat(mat, rect)</c> in OpenCvSharp —
/// like <c>Region</c> in gocv and <c>submat</c> on the JVM — THROWS. So the detectors, which
/// routinely return a box a pixel or two outside the image, produce a working crop in the reference
/// and an exception in a port that translates the slice literally.
/// </para>
///
/// <para>
/// A port that "works" is therefore a port that clamps, and it must clamp the way the slice
/// effectively does. Hence one function, used everywhere, with unit tests per language.
/// </para>
/// </summary>
public static class Crop
{
    /// <summary>
    /// The clamped equivalent of <c>img[y1:y2, x1:x2]</c>.
    ///
    /// <para>
    /// Negative starts clamp to 0 rather than counting from the end. The reference's own coordinates
    /// come from detector output that has already been clipped to non-negative, so the
    /// count-from-the-end branch is unreachable there; implementing it would add behaviour the
    /// reference does not exercise, which is a worse kind of wrong than not implementing it.
    /// </para>
    ///
    /// <para>
    /// An empty intersection yields a zero-sized image rather than an error, because the reference's
    /// slice does too — and the OCR path has a documented degenerate route for exactly that.
    /// </para>
    /// </summary>
    public static Image ClampedCrop(Image src, int x1, int y1, int x2, int y2)
    {
        int w = src.Width, h = src.Height;

        int left = Math.Clamp(x1, 0, w);
        int top = Math.Clamp(y1, 0, h);
        int right = Math.Clamp(x2, 0, w);
        int bottom = Math.Clamp(y2, 0, h);

        // A reversed range is empty in Python, not an error and not a flipped crop.
        int cropW = Math.Max(0, right - left);
        int cropH = Math.Max(0, bottom - top);

        if (cropW == 0 || cropH == 0)
        {
            // Zero-sized, with the source's type so downstream code sees the expected channel count.
            return Image.Wrap(new Mat(cropH, cropW, src.Mat.Type()));
        }

        // Clone rather than return a view: a submat shares the parent's buffer, so the crop would
        // dangle the moment the parent is disposed — and the parent here is a pipeline intermediate
        // that is released as soon as the stage ends.
        using var view = new Mat(src.Mat, new Rect(left, top, cropW, cropH));
        return Image.Wrap(view.Clone());
    }
}

File: ports/go/internal/docproc/imaging/crop.go
package imaging

import (
	"fmt"
	"image"

	"gocv.io/x/gocv"
)

// ClampedCrop is the ONLY sanctioned way to crop in this port. Do not call
// gocv.Mat.Region directly.
//
// This exists because of the highest-risk divergence in the whole port
// (CONVENTIONS §6.6). The Python pipeline crops with a numpy slice:
//
//	img[box[1]:box[3], box[0]:box[2]]     # WordsDetector, TextFieldsDetector
//
// and numpy slicing is FORGIVING in two specific ways that no OpenCV binding is:
//
//   - an upper bound past the edge is silently clamped to the edge;
//   - a NEGATIVE start is interpreted as counting from the end.
//
// The detector clamps negatives to zero only when `resize` is set
// (`detect_res[detect_res < 0] = 0` in postprocessing.py) and never clamps the upper
// bound against the image dimensions at all. So out-of-range boxes DO reach this
// code path in practice.
//
// gocv's Region, OpenCvSharp's Mat[Rect] and the JVM's submat all throw instead. A
// port that "works" is therefore one that clamps — and it must clamp the way numpy
// effectively does, to [0, dim], rather than throwing or returning a differently
// sized crop, or the OCR input silently changes size and the text changes with it.
//
// The negative-start case is deliberately NOT reproduced as from-the-end indexing:
// that behaviour is an accident of numpy rather than intent, it would produce a crop
// from the opposite side of the document, and the detector already zeroes negatives
// on the path that matters. Negatives are clamped to 0 and that choice is recorded
// here rather than left to be rediscovered.
//
// An empty intersection is not an error: OCRv2Preprocessing has an explicit
// degenerate-crop path that returns a 32x16 black patch, so a zero-area box must be
// representable rather than fatal.
func ClampedCrop(src Image, x1, y1, x2, y2 int) (Image, error) {
	w, h := src.Width(), src.Height()

	// Order the corners: a box with x2 < x1 is meaningless, and numpy would produce
	// an empty slice rather than a negative-width one.
	if x2 < x1 {
		x1, x2 = x2, x1
	}
	if y2 < y1 {
		y1, y2 = y2, y1
	}

	x1 = clamp(x1, 0, w)
	x2 = clamp(x2, 0, w)
	y1 = clamp(y1, 0, h)
	y2 = clamp(y2, 0, h)

	if x2 <= x1 || y2 <= y1 {
		// Zero-area after clamping. Return an empty 0x0 image of the right type;
		// callers that need a placeholder patch build one themselves, exactly as
		// OCRv2Preprocessing does.
		return Image{mat: gocv.NewMatWithSize(0, 0, gocv.MatTypeCV8UC3)}, nil
	}

	region := src.mat.Region(image.Rect(x1, y1, x2, y2))
	defer region.Close()
	// Region is a VIEW into the parent. Cloning is mandatory: the parent canvas is
	// reused and rebound across stages, and a view would change under the caller --
	// the same class of bug the service hit by reading PipelineResults after the
	// lease was released.
	return Image{mat: region.Clone()}, nil
}

func clamp(v, lo, hi int) int {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

// CropRect is a convenience wrapper for callers that already hold a rectangle.
func CropRect(src Image, r image.Rectangle) (Image, error) {
	return ClampedCrop(src, r.Min.X, r.Min.Y, r.Max.X, r.Max.Y)
}

// MustBe8UC3 guards the assumption most of this package makes.
func MustBe8UC3(i Image) error {
	if i.mat.Type() != gocv.MatTypeCV8UC3 {
		return fmt.Errorf("imaging: expected CV_8UC3, got %v", i.mat.Type())
	}
	return nil
}
